content/en/blog/general/why-chainsaw-is-unique/index.md

```diff
@@ -0,0 +1,200 @@
+
+Let's say we want an assertion to consider only containers that have no security context defined.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    k8s-app: kube-dns
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      (containers[?securityContext == null]):
+        (len(@)): 0
+```
+
+In the assertion above, the first three fields `spec`, `template`, and `spec` are basic projections that simply take the content of their respective fields and pass that to descendants.
+
+`(containers[?securityContext == null])` is a JMESPath expression filtering the `containers` array, selecting only the element where `securityContext` is `null`. This projection results in a new array that is passed to the descendant (`(len(@))` in this case).
+
+Finally `(len(@))` is reached, it's another JMESPath expression that computes the length of the array. There's no more descendant at this point, we're of a leaf of the YAML tree, and the array length we just computed is then compared against 0.
+
+If the comparison matches, the assertion will be considered valid, if not it will be considered failed.
+
+In simple words, the assertion above checks that all containers in the Deployment have a security context configured.
+
+### Iterating over arrays (or maps)
+
+Being able to filter arrays allows selecting the elements to be processed.
+
+On top of that Chainsaw allows iterating over array elements to validate each item separately.
+
+The assertion above can be rewritten using an iterator:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    k8s-app: kube-dns
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      # the ~ modifier tells Chainsaw to iterate over the array elements
+      ~.(containers):
+        securityContext: {}
+```
+
+This assertion uses the `~` modifier and Chainsaw will evaluate descendants once per element in the array.
+
+### Explicit bindings and reference
+
+When passing from a parent node to descendant nodes, it can be useful to keep a reference to one or more parent nodes.
+
+Let's say we want to consider the pod level security context too, we can write such an assertion with the syntax below:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    k8s-app: kube-dns
+  namespace: kube-system
+spec:
+  template:
+    (spec)->specBinding:
+      # the ~ modifier tells Chainsaw to iterate over the array elements
+      ~.(containers):
+        ($specBinding.securityContext != null || securityContext != null): true
+```
+
+In the assertion above, `(spec)->specBinding` means that Chainsaw will keep a reference to the result of the expression evaluation and will make it available to descendants under the `$specBinding` binding.
+
+Now we can use the binding to write the assertion check taking the pod level security context into account like this `($specBinding.securityContext != null || securityContext != null): true`.
```

Comment:
This is complex. You should write at least one sentence describing, in plain English, what this assertion does. Based on studying the example and what you wrote, it's something like: "This assertion checks to ensure that, if any container's securityContext is not defined, the Pod-level securityContext is defined."

Reply:
Done, thanks.